File: backend/app/schemas/curator.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import BookingStatus, ComplaintStatus, Currency, Gender


class StudentListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    phone: str
    first_name: str | None = None
    last_name: str | None = None
    avatar_url: str | None = None
    is_blocked: bool
    is_verified: bool
    created_at: datetime

    # Student profile
    hemis_id: str | None = None
    university_id: int | None = None
    university_name: str | None = None
    university_short: str | None = None
    faculty: str | None = None
    course: int | None = None
    group_name: str | None = None

    # Live stats
    active_bookings_count: int = 0
    total_bookings_count: int = 0
    open_complaints_count: int = 0


class StudentListFilter(BaseModel):
    university_id: int | None = None
    course: int | None = None
    is_blocked: bool | None = None
    q: str | None = None
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)


class LandlordListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    phone: str
    first_name: str | None = None
    last_name: str | None = None
    avatar_url: str | None = None
    email: str | None = None
    is_blocked: bool
    is_verified: bool
    created_at: datetime

    is_pro: bool = False
    is_verified_landlord: bool = False
    houses_count: int = 0
    active_bookings_count: int = 0
    open_complaints_count: int = 0


class LandlordListFilter(BaseModel):
    is_pro: bool | None = None
    is_verified_landlord: bool | None = None
    is_blocked: bool | None = None
    q: str | None = None
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)


class StudentBookingSummary(BaseModel):
    id: int
    house_id: int
    house_title: str | None = None
    house_region: str | None = None
    start_date: date
    end_date: date
    monthly_price: Decimal
    currency: Currency
    status: BookingStatus
    created_at: datetime


class StudentComplaintSummary(BaseModel):
    id: int
    subject: str
    status: ComplaintStatus
    created_at: datetime
    resolved_at: datetime | None = None


class CurrentLandlordInfo(BaseModel):
    """Landlord of the student's currently active rental."""
    id: int
    first_name: str | None = None
    last_name: str | None = None
    phone: str
    avatar_url: str | None = None
    is_pro: bool = False
    house_id: int
    house_title: str | None = None
    booking_id: int
    booking_start_date: date
    booking_end_date: date


class StudentDetail(BaseModel):
    id: int
    phone: str
    first_name: str | None = None
    last_name: str | None = None
    email: str | None = None
    avatar_url: str | None = None
    is_blocked: bool
    is_verified: bool
    last_login_at: datetime | None = None
    created_at: datetime

    # Academic
    hemis_id: str | None = None
    university_id: int | None = None
    university_name: str | None = None
    university_short: str | None = None
    faculty: str | None = None
    course: int | None = None
    group_name: str | None = None
    gender: Gender | None = None
    birth_date: date | None = None

    # Stats
    active_bookings_count: int = 0
    total_bookings_count: int = 0
    open_complaints_count: int = 0
    total_complaints_count: int = 0
    total_spent: Decimal = Decimal(0)

    bookings: list[StudentBookingSummary] = []
    complaints: list[StudentComplaintSummary] = []
    current_landlord: CurrentLandlordInfo | None = None

File: frontend/src/lib/api/curator.ts
import { apiClient } from "./client";
import type { BookingStatus, ComplaintStatus, PaginatedResponse } from "./types";

export interface StudentListItem {
  id: number;
  phone: string;
  first_name?: string | null;
  last_name?: string | null;
  avatar_url?: string | null;
  is_blocked: boolean;
  is_verified: boolean;
  created_at: string;
  hemis_id?: string | null;
  university_id?: number | null;
  university_name?: string | null;
  university_short?: string | null;
  faculty?: string | null;
  course?: number | null;
  group_name?: string | null;
  active_bookings_count: number;
  total_bookings_count: number;
  open_complaints_count: number;
}

export interface StudentListFilter {
  university_id?: number;
  course?: number;
  is_blocked?: boolean;
  q?: string;
  page?: number;
  page_size?: number;
}

export interface StudentBookingSummary {
  id: number;
  house_id: number;
  house_title?: string | null;
  house_region?: string | null;
  start_date: string;
  end_date: string;
  monthly_price: string;
  currency: string;
  status: BookingStatus;
  created_at: string;
}

export interface StudentComplaintSummary {
  id: number;
  subject: string;
  status: ComplaintStatus;
  created_at: string;
  resolved_at?: string | null;
}

export interface CurrentLandlordInfo {
  id: number;
  first_name?: string | null;
  last_name?: string | null;
  phone: string;
  avatar_url?: string | null;
  is_pro: boolean;
  house_id: number;
  house_title?: string | null;
  booking_id: number;
  booking_start_date: string;
  booking_end_date: string;
}

export interface StudentDetail {
  id: number;
  phone: string;
  first_name?: string | null;
  last_name?: string | null;
  email?: string | null;
  avatar_url?: string | null;
  is_blocked: boolean;
  is_verified: boolean;
  last_login_at?: string | null;
  created_at: string;

  hemis_id?: string | null;
  university_id?: number | null;
  university_name?: string | null;
  university_short?: string | null;
  faculty?: string | null;
  course?: number | null;
  group_name?: string | null;
  gender?: "male" | "female" | null;
  birth_date?: string | null;

  active_bookings_count: number;
  total_bookings_count: number;
  open_complaints_count: number;
  total_complaints_count: number;
  total_spent: string;

  bookings: StudentBookingSummary[];
  complaints: StudentComplaintSummary[];
  current_landlord?: CurrentLandlordInfo | null;
}

function buildParams(filters: Record<string, unknown>): URLSearchParams {
  const p = new URLSearchParams();
  for (const [k, v] of Object.entries(filters)) {
    if (v !== undefined && v !== null && v !== "") p.append(k, String(v));
  }
  return p;
}

export interface LandlordListItem {
  id: number;
  phone: string;
  first_name?: string | null;
  last_name?: string | null;
  avatar_url?: string | null;
  email?: string | null;
  is_blocked: boolean;
  is_verified: boolean;
  created_at: string;
  is_pro: boolean;
  is_verified_landlord: boolean;
  houses_count: number;
  active_bookings_count: number;
  open_complaints_count: number;
}

export interface LandlordListFilter {
  is_pro?: boolean;
  is_verified_landlord?: boolean;
  is_blocked?: boolean;
  q?: string;
  page?: number;
  page_size?: number;
}

export const curatorApi = {
  listStudents: async (filters: StudentListFilter = {}) => {
    const params = buildParams(filters);
    const res = await apiClient.get<PaginatedResponse<StudentListItem>>(
      `/curator/students?${params.toString()}`,
    );
    return res.data;
  },

  getStudent: async (id: number) => {
    const res = await apiClient.get<StudentDetail>(`/curator/students/${id}`);
    return res.data;
  },

  listLandlords: async (filters: LandlordListFilter = {}) => {
    const params = buildParams(filters as Record<string, unknown>);
    const res = await apiClient.get<PaginatedResponse<LandlordListItem>>(
      `/curator/landlords?${params.toString()}`,
    );
    return res.data;
  },
};
